packages/langium/src/workspace/documents.ts

```diff
@@ -177,21 +191,37 @@ export class DefaultLangiumDocumentFactory implements LangiumDocumentFactory {
         if (uri === undefined) {
             uri = URI.parse(textDocument!.uri);
         }
-        const services = this.serviceRegistry.getServices(uri);
+
         let parseResult: ParseResult<T>;
         if (model === undefined) {
-            parseResult = services.parser.LangiumParser.parse<T>(text ?? textDocument!.getText());
+            parseResult = this.parse<T>(uri, text ?? textDocument!.getText());
         } else {
             parseResult = { value: model, parserErrors: [], lexerErrors: [] };
         }
+
+        const servicesReqistry = this.serviceRegistry;
         return this.createLangiumDocument<T>(parseResult, uri, textDocument ?? {
             $template: true,
-            languageId: services.LanguageMetaData.languageId,
+            languageId: servicesReqistry.getServices(uri).LanguageMetaData.languageId,
             uri,
             text
         });
     }
 
+    update<T extends AstNode = AstNode>(document: LangiumDocument<T>): LangiumDocument<T> {
+        const textDocument = this.textDocuments.get(document.uri.toString()) ?? document.textDocument;
```

Comment:
Suggestion: I believe we should never re-use the existing `document.textDocument` here, since it will always be outdated. We should resort to asking the file system for the text instead, if the document is not in the list of text documents.

Reply:
You're right. I just checked the case of a change event being triggered from outside with no editor being opened in VS code. I that case it's definitely outdated.